### ISSUE TITLE
Override simple selectors in @use

### ISSUE THREAD

**Issue by nategreen** [sass/sass#2871]
**TL;DR:** I think I'd like to be able to override simple selectors from a module that outputs CSS, the same way I can override variables. (Or, if there's an existing relatively-elegant solution to my issue below, I'd welcome that advice instead.)

I've been trying to come up with a simple way to create styles that can easily be used in projects with various stacks (Angular, React, some jQuery, some Vanilla JS…).

One thing I often want to do is override selectors downstream. So, for instance, if I have a class `.button` that I want to use in a web component, I need to `@extend .button` to get the same styles applied to `:host`. (And it'll create extraneous selectors e.g. `:host, .button`—which I'm not _too_ concerned about, but it is a mild nuisance when we're trying to keep filesize down.)

I'm comfortable writing complicated SCSS (maybe _too_ comfortable?) but I want to keep the styles as straightforward as I can, so that contributors from other internal teams can easily follow existing examples and write good CSS for new components.

**Example 1**
```scss
// _button.scss
$_selectors: (
  default: .button,
  hover: :hover, .button--state-hover
);

@mixin button( $include: $_selectors ) {
  @if map-has-key($include, default) {
    #{map-get($include, default)} {
      appearance: none;
      background: red;
      color: white;
    }
  }
  @if map-has-key($include, hover) {
    #{map-get($include, hover)} {
      background: pink;
    }
  }
}

// downstream packages
@use "~package/scss/button" as *;

@include button((
  default: :host,
  hover: :host:hover, :host.button--state-hover
));

// I've also done it to where you'd override selectors declared as individual variables;
// and maybe one or two other variations on that theme.
```

One issue with that method is that for a moderately complex component like buttons (they have loading states, different sizes, different widths, icons on one or both sides…) the mixin gets very large and difficult to navigate with all that conditional stuff going on. And it's especially hard to read when you have something like `dark theme > primary button > hover`.

**Example 2**
```scss
// _button.scss
@mixin button {
  appearance: none;
  background: red;
  color: white;
}

@mixin button--hover {
  background: pink;
}

// downstream
@use "~package/scss/button" as *;

:host {
  @include button;
  
  &:hover {
    @include button--hover;
  }
}
```

While considerably simpler, this method requires downstream packages to keep in step with the states/properties of each component. Furthermore, it doesn't really work well for combined states, and I certainly don't want to be writing `button--dark-primary-hover`.

**Example 3**
(This is just what I mentioned above…writing BEM-style classes and/or attributes and expecting downstream packages to `@extend` conservatively.)

```scss
// _button.scss
.button {
  appearance: none;
  background: red;
  color: white;
}

.button:hover, .button--state-hover {
  background: pink;
}

// downstream
@use "~package/scss/button" as *;

:host {
  @extend .button;
}
```

**What I'd like to do**

Since the main thing I'm trying to overcome is replacing single selectors, I'm proposing the following. I stuck with "simple selectors" because I expect that's probably easier to reckon with than somehow replacing complex selectors or introducing a whole new _thing_ to somehow name complex selectors. (So far, most complex selectors come up because we're trying to provide support for doing the same thing in a couple of different ways, e.g. classes or attributes.)

```scss
// _button.scss
.button {
  appearance: none;
  background: red;
  color: white;
}

.button:hover, .button--state-hover {
  background: pink;
}

// downstream
@use 'package/scss/button' with (
  .button: :host
);

// output
:host {
  appearance: none;
  background: red;
  color: white;
}

:host:hover, .button--state-hover {
  background: pink;
}
```

Maybe in the same mechanism, we could include/exclude selectors as well, maybe by setting `.button--state-hover` to `null`. That might be overreaching, though. 😅

**Comment by nategreen** [comment]
Oh, looks like this might overlap with #2566 and maybe #2791? And that reminds me that another approach I tried and rejected was a placeholder-based approach (which sort of spawned #2570).

Scoped `@extend` might work for this? Especially if we could prevent _the rest of_ the CSS from being passed along automatically. So, I don't think the syntax proposed in #2791 would help with that, but maybe if it was something like a `!flag` that could be thrown to prevent styles from being automatically included, couples with the scoped `@extend`…it would be like effectively treating simple selectors as placeholders (I think).

```scss
// downstream
@use "package/scss/button' as button !no-styles;

:host {
  @extend .button from button;
}
```

**Comment by jathak** [comment]
Sass generally avoids optimizing for the use case of modifying the behavior of upstream dependencies in ways that aren't explicitly allowed by those files.

For dependencies you control, the recommended method of allowing configuration is through mixins. `!default` variable configuration is also an option, though support for that was added to the module system more for backwards compatibility than anything else.

If you want to configure a third-party dependency, you should file an issue against the project in question asking them to add support for that through a mixin or `!default` variable. We don't like to add features to Sass that allow arbitrary configuration of dependencies so that library authors don't have to worry about breaking their downstream users if they make a change to their implementation that doesn't change their public API.
